### PR TITLE
Pinpointer refactor

### PR DESCRIPTION
**Pull request by ghost**
:cl: Xhuis
add: Nanotrasen electricians have introduced a software update to all licensed pinpointers. The new pinpointers are more responsive and automatically switch modes based on the current crisis.
add: It should be noted that some of these pinpointers have been reported as missing from their distribution warehouse and are presumed to have been stolen. This is likely nothing to worry about.
/:cl:
- Pinpointers now use the `fastprocess` subsystem instead of a `spawn` chain.
- Pinpointers now have messages and sounds when turning them on or off.
- Pinpointers now dynamically switch tracking modes based on what's necessary. For instance, when tracking a malfunctioning AI, the pinpointers will swap between locating the AI core and the AI's APC. depending on which is currently occupied.
  - If tracking is finicky, turning the pinpointer on and off resets its target so it can search again.
  - Pinpointers also lose their target every five seconds to give them an opportunity to update.
- Pinpointers now have a different appearance and play a sound whenever a nuke is armed.
  - Syndicate pinpointers track the infiltrator once this occurs.
- The advanced pinpointer has been removed in favor of two tracking modes that can be accomplished through varediting.

### CODE REVIEW

File: code/game/gamemodes/nuclear/pinpointer.dm [hunk 1]
```diff
@@ -21,277 +24,117 @@
...
+					target = A
+		if("infiltrator") //We track the Syndicate infiltrator, so we can find our way back
+			target = SSshuttle.getShuttle("syndicate")
+		if("operative")
+			var/list/possible_targets = list()
+			for(var/datum/mind/M in ticker.mode.syndicates)
+				if(M.current && M.current.stat != DEAD && M.current.client)
+					possible_targets |= M.current
+			if(possible_targets.len)
+				target = pick(possible_targets)
```
Comment (coiax) [inline code/game/gamemodes/nuclear/pinpointer.dm]: Why not just work out which of the targets is closer with get_dist?

File: code/game/gamemodes/nuclear/pinpointer.dm [hunk 2]
```diff
@@ -8,9 +9,11 @@
...
 	throw_speed = 3
 	throw_range = 7
-	materials = list(MAT_METAL=500)
-	var/obj/item/weapon/disk/nuclear/the_disk = null
-	var/active = 0
+	materials = list(MAT_METAL = 500, MAT_GLASS = 250)
+	var/active = FALSE
+	var/atom/movable/target = null //The thing we're searching for
+	var/nuke_warning = FALSE // If we've set off a miniature alarm about an armed nuke
+	var/mode = "nuclear" //What are we looking for?
```
Comment (RemieRichards) [inline code/game/gamemodes/nuclear/pinpointer.dm]: you already know what's wrong with this.

File: code/game/gamemodes/nuclear/pinpointer.dm [hunk 3]
```diff
@@ -21,277 +24,117 @@
...
-	spawn(5) .()
-
-
-/obj/item/weapon/pinpointer/nukeop/proc/worklocation()
+	switch(mode)
+		if("nuclear") //We track the nuke disk, for protection or more nefarious purposes
+			var/obj/item/weapon/disk/nuclear/N = locate()
+			target = N
+		if("malf_ai") //We track the AI, who wants to blow us all to smithereens
+			for(var/mob/living/silicon/ai/A in ai_list)
```
Comment (RemieRichards) [inline code/game/gamemodes/nuclear/pinpointer.dm]: this should be an istypeless forloop.

File: code/game/gamemodes/nuclear/pinpointer.dm [hunk 4]
```diff
@@ -21,277 +24,117 @@
...
+				if(A.nuking)
+					target = A
+			for(var/obj/machinery/power/apc/A in apcs_list)
```
Comment (RemieRichards) [inline code/game/gamemodes/nuclear/pinpointer.dm]: this should be an istypeless forloop.

File: code/game/gamemodes/nuclear/pinpointer.dm [hunk 5]
```diff
@@ -21,277 +24,117 @@
...
+				if(A.malfhack && A.occupier)
+					target = A
+		if("infiltrator") //We track the Syndicate infiltrator, so we can find our way back
+			target = SSshuttle.getShuttle("syndicate")
+		if("operative")
+			var/list/possible_targets = list()
+			for(var/datum/mind/M in ticker.mode.syndicates)
```
Comment (RemieRichards) [inline code/game/gamemodes/nuclear/pinpointer.dm]: this should be an istypeless forloop.

File: code/game/gamemodes/nuclear/pinpointer.dm (hunk 2 above)
Comment (Xhuis) [inline code/game/gamemodes/nuclear/pinpointer.dm]: I actually don't. Should I be using defines?

Comment (RemieRichards) [inline code/game/gamemodes/nuclear/pinpointer.dm]: Well you just said it, so you clearly do~

**Comment by coiax** [comment]
That's just Remie, they're just a harsh critquer, don't take it personally.

**Comment by RemieRichards** [comment]
```>doing my job as a maintainer
>bashing```

Really?

**Comment by Xhuis** [comment]
Yeah. Instead of saying "hey, you should use defines", you said "you know what's wrong with this." then made a smug comment when I asked if that was what you wanted me to do.

**Comment by RemieRichards** [comment]
Holy SHIT your negative bias towards maintainers /really/ fucking shines through. You think I was being mean and smug before? fucking buckle up buddy.

**Comment by coiax** [comment]
let's not get all Niknak/coiax in here

File: code/game/gamemodes/nuclear/pinpointer.dm (hunk 1 above)
Comment (Xhuis) [inline code/game/gamemodes/nuclear/pinpointer.dm]: Trying my best to do this as I type this comment out, but Byond doesn't like associative lists and is difficult. If you have any advice, I'd love to hear it.

File: code/game/gamemodes/nuclear/pinpointer.dm [hunk 6]
```diff
@@ -8,290 +9,148 @@
...
+		if(TRACK_INFILTRATOR)
+			target = SSshuttle.getShuttle("syndicate")
+		if(TRACK_OPERATIVES)
+			var/list/possible_targets = list()
+			var/turf/here = get_turf(src)
+			for(var/V in ticker.mode.syndicates)
+				var/datum/mind/M = V
+				if(M.current && M.current.stat != DEAD)
+					possible_targets |= M.current
+			var/mob/living/closest_operative
```
Comment (RemieRichards) [inline code/game/gamemodes/nuclear/pinpointer.dm]: ```var/mob/living/closest_operative = get_closest_atom(/mob/living/carbon/human, possible_targets, here)```
Everything below besides
```
if(closest_operative)
    target = closest_operative
```
is wrong.

**Comment by Shadowlight213** [comment]
So does this remove the advanced pinpointer, as well as the ability to track anything except the disk/malf ai/ops? Should put that in the changelog.

From reading the code and your comments it looks like the pinpointer can't track anything that's not one of the hardcoded modes.
Major :-1:

File: code/game/gamemodes/nuclear/pinpointer.dm [hunk 7]
```diff
@@ -8,290 +9,140 @@
...
+		target = null //Restarting the pinpointer forces a target reset
+		STOP_PROCESSING(SSfastprocess, src)
 
 /obj/item/weapon/pinpointer/examine(mob/user)
 	..()
+	switch(mode)
+		if(TRACK_NUKE_DISK)
+			user << "Its tracking indicator reads \"nuclear_disk\"."
+		if(TRACK_MALF_AI)
+			user << "Its tracking indicator reads \"01000001 01001001\"."
```
Comment (GunHog) [inline code/game/gamemodes/nuclear/pinpointer.dm]: Lol it means "AI".

File: code/game/gamemodes/nuclear/pinpointer.dm [hunk 8]
```diff
@@ -8,290 +9,140 @@
...
+		if(TRACK_INFILTRATOR)
+			user << "Its tracking indicator reads \"vasvygengbefuvc\"."
```
Comment (GunHog) [inline code/game/gamemodes/nuclear/pinpointer.dm]: Rot13 for "infiltratorship", even though it is Syndicate, not Steampunk. Still, nice little touch.

File: code/game/gamemodes/nuclear/pinpointer.dm [hunk 9]
```diff
@@ -8,290 +9,140 @@
...
+	target = null //Switch modes so we can find the new target
+
+/obj/item/weapon/pinpointer/proc/refresh_target() //Periodically removes the target to allow the pinpointer to update (i.e. malf AI shunts, an operative dies)
+	target = null
+
+/obj/item/weapon/pinpointer/syndicate //Syndicate pinpointers automatically point towards the infiltrator once the nuke is active.
+	name = "syndicate pinpointer"
+	desc = "A handheld tracking device that locks onto certain signals. It's configured to switch tracking modes once it detects the activation signal of a nuclear device."
+
+/obj/item/weapon/pinpointer/syndicate/cyborg //Cyborg pinpointers just look for a random operative.
```
Comment (GunHog) [inline code/game/gamemodes/nuclear/pinpointer.dm]: Would it be possible to allow the borg to cycle through the Ops using attack_self?

File: code/game/gamemodes/nuclear/pinpointer.dm (hunk 8 above)
Comment (Xhuis) [inline code/game/gamemodes/nuclear/pinpointer.dm]: That's supposed to represent just garbling it so people who aren't Syndicate can't read it.

File: code/game/gamemodes/nuclear/pinpointer.dm (hunk 9 above)
Comment (Xhuis) [inline code/game/gamemodes/nuclear/pinpointer.dm]: Possibly, but it might be more trouble than it's worth.

**Comment by Xhuis** [comment]
@Shadowlight213 It does remove the advanced pinpointer - something I didn't mention since it doesn't ever see use. Nonetheless, this rewrite is almost mandatory because of how terrible the code is. I _guess_ I could see about customized tracking but it might be difficult.

**Comment by GunHog** [comment]
Please make it somewhat easy for me to make a Xeno Queen pinpointer for @WJohn

**Comment by kevinz000** [comment]
:-1: if adv pinpointer is removed.
It's not useless.

**Comment by Xhuis** [comment]
@GunHog It's already very simple with this PR.
@kevinz000 I was under the assumption that the advanced pinpointer wasn't even used. If wanted, I could easily re-add it using the new code.

**Comment by optimumtact** [comment]
@Shadowlight213 @kevinz000 explain what you used the advanced pinpointer for and why you need the pinpointer to point to other items otherwise there is no valid reason to hold the PR up.

Thanks

**Comment by coiax** [comment]
Having the pinpointer point to other items is standard admin var editing for events.

**Comment by Xhuis** [comment]
Although I've never once seen that done, I'll humor you and cook something up.

**Comment by Xhuis** [comment]
I added two new tracking modes - `TRACK_ATOM` and `TRACK_COORDINATES`, settable by changing `mode` to 5 and 6 respectively. 

`TRACK_ATOM` constantly looks for a particular object, designated by the `constant_target` variable; it can be changed by hitting the pinpointer with another object at any time to make the pinpointer point towards that object instead.

`TRACK_COORDINATES` points towards a location with the specified X and Y coordinates (`target_x` and `target_y`) on the pinpointer's z-level.

File: code/game/gamemodes/malfunction/Malf_Modules.dm [hunk 10]
```diff
@@ -40,10 +40,8 @@
 	doomsday_device = DOOM
 	doomsday_device.start()
 	verbs -= /mob/living/silicon/ai/proc/nuke_station
-	for(var/obj/item/weapon/pinpointer/point in pinpointer_list)
-		for(var/mob/living/silicon/ai/A in ai_list)
-			if((A.stat != DEAD) && A.nuking)
-				point.the_disk = A //The pinpointer now tracks the AI core
+	for(var/obj/item/weapon/pinpointer/P in pinpointer_list)
```
Comment (GunHog) [inline code/game/gamemodes/malfunction/Malf_Modules.dm]: RR would say something about "istypless" for loops here.

File: code/game/gamemodes/nuclear/pinpointer.dm [hunk 11]
```diff
@@ -8,290 +10,162 @@
...
-			user << "No operatives detected within scanning range."
+/obj/item/weapon/pinpointer/proc/my_god_jc_a_bomb() //If we should get the hell back to the ship
+	for(var/obj/machinery/nuclearbomb/bomb in machines)
+		if(bomb.timing)
+			if(!nuke_warning)
+				nuke_warning = TRUE
+				playsound(src, 'sound/items/Nuke_toy_lowpower.ogg', 50, 0)
+				if(isliving(loc))
+					var/mob/living/L = loc
+					L << "<span class='userdanger'>Your [name] vibrates and lets out a tinny alarm. Uh oh.</span>"
```
Comment (GunHog) [inline code/game/gamemodes/nuclear/pinpointer.dm]: tiny

Comment (Xhuis) [inline code/game/gamemodes/nuclear/pinpointer.dm]: Nah, that's intentional. If you listen to that sound, it's really low-quality. I'll upload it in a sec.

File: code/game/gamemodes/malfunction/Malf_Modules.dm (hunk 10 above)
Comment (Xhuis) [inline code/game/gamemodes/malfunction/Malf_Modules.dm]: Why, though? The only things in the list are pinpointers. I don't know how to optimize fors/lists easily so let me know if I need to change it.

Comment (RemieRichards) [inline code/game/gamemodes/malfunction/Malf_Modules.dm]: that's precisely why it should be an istypeless loop.

**Comment by coiax** [comment]
Xhuis, what I don't understand is why we have this refresh every 5 second mode which overrides the tracked object atom, surely we'd just follow the atom and rescan if it's qdelled or made null

File: code/game/gamemodes/malfunction/Malf_Modules.dm (hunk 10 above)
Comment (Xhuis) [inline code/game/gamemodes/malfunction/Malf_Modules.dm]: ...oh. Ok.

Comment (RemieRichards) [inline code/game/gamemodes/malfunction/Malf_Modules.dm]: the idea is that if we know the contents of the list we don't want byond ```istype()```ing it every for iteration (it's probably not a real ```istype()``` likely it's an internal one in C++ but it still adds up)

**Comment by Xhuis** [comment]
The refreshing is designed to ensure that the pinpointer doesn't "stick" to its target - for instance, if it's tracking a shunted APC and the AI shunts back, it'll still track the APC even though the AI isn't there anymore. The five-second refresh is so that the pinpointer can make sure that it doesn't need to track a new target.

Varediting the pinpointer to do what it used to is extremely simple. Just change the mode to "5" (to track an atom) and set `constant_target` to the atom you want to track.

**Comment by optimumtact** [comment]
Shit forgot to merge this

**Comment by Xhuis** [comment]
You da man :heart:

**Pull request merged**